core/src/main/java/com/github/games647/fastlogin/core/storage/MigrationManager.java

```diff
@@ -0,0 +1,124 @@
+package com.github.games647.fastlogin.core.storage;
+
+import com.github.games647.fastlogin.core.shared.FastLoginCore;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+public class MigrationManager {
+
+    protected static final String MIGRATION_TABLE = "migrations";
+    protected static final String CREATE_TABLE_STMT = "CREATE TABLE IF NOT EXISTS `" + MIGRATION_TABLE + "` ("
+            + "`ID` INTEGER PRIMARY KEY AUTO_INCREMENT, "
+            + "`Table` VARCHAR(32), "
+            + "`Version` INTEGER NOT NULL, "
+            + "`Date` TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP "
+            + ')';
+
+    protected static final String LOAD_TABLE_VERSION = "SELECT MAX(Version) FROM `" + MIGRATION_TABLE
+            + "` WHERE `Table`=? LIMIT 1";
+
+    protected static final String INSERT_MIGRATION = "INSERT INTO `" + MIGRATION_TABLE
+            + "` (`Table`, `Version`) " + "VALUES (?, ?) ";
+
+    protected final FastLoginCore<?, ?, ?> core;
+
+    protected final SQLStorage storage;
+
+    protected MigrationManager(FastLoginCore<?, ?, ?> core, SQLStorage storage) {
+        this.core = core;
+        this.storage = storage;
+    }
+
+    protected void createTables() throws SQLException {
+        try (Connection con = storage.getDataSource().getConnection();
+                Statement createStmt = con.createStatement()) {
+
+            // if (dataSource.getDriverClassName().contains("sqlite")) {
+            // throws: the return value of "fastlogin.hikari.HikariDataSource.getDriverClassName()" is null
+            if (storage instanceof SQLiteStorage) {
+                createStmt.executeUpdate(CREATE_TABLE_STMT.replace("AUTO_INCREMENT", "AUTOINCREMENT"));
+            } else {
+                createStmt.executeUpdate(CREATE_TABLE_STMT);
+            }
+
+        }
+    }
+
+    protected int getTableVersion(String table) {
+        try (Connection con = storage.getDataSource().getConnection();
+                PreparedStatement loadStmt = con.prepareStatement(LOAD_TABLE_VERSION)) {
+            loadStmt.setString(1, table);
+
+            try (ResultSet resultSet = loadStmt.executeQuery()) {
+                if (resultSet.next()) {
+                    return resultSet.getInt(1);
+                }
+            }
+        } catch (SQLException sqlEx) {
+            core.getPlugin().getLog().error("Failed to query version of table {}", table, sqlEx);
+        }
+
+        return -1;
+    }
+
+    protected void migrateTable(MigratableStorage table) {
+        int initialVersion = getTableVersion(table.getTableName());
+
+        // special case: table premium was created before the migration manager
+        if (initialVersion == 0 && "premium".equals(table.getTableName()) && table.tableExists()) {
+            initialVersion = 1;
+        }
+
+        for (int i = initialVersion; i < table.getLatestTableVersion(); i++) {
```

Comment:
We should better differentiate between what `latest` or `getTableVersion` means. Maybe something like `getCurrentVersion` and `getRequiredVersion`. Or add a comment that makes it clear.

Reply:
Changed in e401103bc93a0d365499887672a99b46a3bcb752